### PR TITLE
Add Iter.map_while

### PR DESCRIPTION
**Pull request by OlivierNicole**
I found myself needing this function and didn’t find an easy way to make it in an efficient way by composing simpler functions.

### CODE REVIEW

**Comment by c-cube** [comment]
This seems quite useful indeed. I do have one design question: what if you want to stop before yielding any element (e.g. if the first element is "wrong")?

Would this be more general?
```ocaml
val map_while : ('a -> [`Yield of 'b | `Return of 'b | `Stop]) -> 'a t -> 'b t
```

where yield means, yield and continue, and return means yield and stop iterating

**Comment by OlivierNicole** [comment]
It does seem more general.

**Comment by OlivierNicole** [comment]
I implemented this version.
Is there a requirement to be compatible with OCaml 4.08? The CI fails for this reason.

**Comment by c-cube** [comment]
Yes, I'd like to preserve compatibility with 4.08 for now. You can use `CCList.equal` instead.

File: tests/unit/t_iter.ml [hunk 1]
```diff
@@ -237,6 +237,22 @@ let () =
...
+  OUnit.assert_equal
+    ~cmp:(CCList.equal Int.equal)
+    (1 -- 10
+     |> map_while (fun x -> if x = 7 then `Return (x + 1) else `Yield (x - 1))
+     |> to_list)
+    [0; 1; 2; 3; 4; 5; 8]
+
+let () =
+  OUnit.assert_equal
+    ~cmp:(List.equal Int.equal)
```
Comment (c-cube) [inline tests/unit/t_iter.ml:250]:
```suggestion
    ~cmp:(CCList.equal Int.equal)
```

**Comment by c-cube** [comment]
I'll merge and fix the tests. Thank you :)

**Pull request merged**